src/shared/hostname-setup.c

```diff
@@ -238,6 +248,67 @@ static const char* const hostname_source_table[] = {
 
 DEFINE_STRING_TABLE_LOOKUP(hostname_source, HostnameSource);
 
+int hostname_substitute_wildcards(char *name) {
+        static const sd_id128_t key = SD_ID128_MAKE(98,10,ad,df,8d,7d,4f,b5,89,1b,4b,56,ac,c2,26,8f);
+        sd_id128_t mid = SD_ID128_NULL;
+        size_t left_bits = 0, counter = 0;
+        uint64_t h = 0;
+        int r;
+
+        assert(name);
+
+        /* Replaces every occurence of '?' in the specified string with a byte hashed from
+         * /etc/machine-id. This is supposed to be used on /etc/hostname files that want to automatically
+         * configure a hostname derived from the machine ID in some form.
+         *
+         * Note that this does not directly use the machine ID, because that's not necessarily supposed to be
+         * public information to be broadcast on the network, while the hostname certainly is. */
+
+        for (char *n = name; *n; n++) {
+                if (*n != '?')
+                        continue;
+
+                if (left_bits <= 0) {
+                        if (sd_id128_is_null(mid)) {
+                                r = sd_id128_get_machine(&mid);
+                                if (r < 0)
+                                        return r;
+                        }
+
+                        struct siphash state;
+                        siphash24_init(&state, key.bytes);
+                        siphash24_compress(&mid, sizeof(mid), &state);
+                        siphash24_compress(&counter, sizeof(counter), &state); /* counter mode */
+                        h = siphash24_finalize(&state);
+                        left_bits = sizeof(h) * 8;
+                        counter++;
+                }
+
+                assert(left_bits >= 4);
+                *n = hexchar(h);
```

Comment:
Not a matter, but I'd like to pass `hexchar(h & 0x0f)` to not pass too large value, as `hexchar()` takes `int`.
Just a paranoia, though.

Reply:
valid paranoia, but none that hexchar() already does the same paranoia mangling internally ;-)